### ISSUE TITLE
[Metal Trace] tt_eager test hang without dprint

### ISSUE THREAD

**Issue by tooniz** [tenstorrent/tt-metal#7793]
**Describe the bug**
I've implemented the device mode capture for trace, which can be easily integrated into tt eager tests.

The smallest test where I could reproduce the hang is on the following, it hangs in debug/release mode but with drprint enabled the hang disappears. 

```
commit 930d42694bcf720702226d5d8373b2c64fd7ec88 (HEAD -> tzhou/fd2-trace-eager, origin/tzhou/fd2-trace-eager)
Author: Tony Zhou <tzhou@tenstorrent.com>
Date:   Tue Apr 23 15:47:24 2024 +0000

    #0: tt_eager pooling test with trace, turning off DPRINT makes one of the tests fail.
    
    WHOLE SUITE PASSES:
    - `TT_METAL_DPRINT_CORES=all pytest "tests/tt_eager/python_api_testing/unit_testing/misc/test_average_pool.py::test_run_average_pool"`
    
    FAIL:
    - `pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_average_pool.py::test_run_average_pool[BFLOAT16-resnet50_unpadded]"`
```

**Comment by tooniz** [comment]
Let's split the effort, I found another bug at tt_eager level related to trace buffer corruption. https://github.com/tenstorrent/tt-metal/issues/7794 I'll take that one

**Comment by kmabeeTT** [comment]
Debugged this (found a set of DPRINTS in prefetcher that can have hang persist) and I see we hang in Prefetcher `fetch_q_get_cmds()` where it finds no more work after executing the Trace via `ExecBuf` cmd (and a few following `RelayInline` cmds), and it is looking like a restriction is missed. 

> Is the exec_buf command in its own FetchQ entry?  That’s a requirement.  If not, that would explain things…

Confirmed with @tooniz that nothing was explicitly addded for this constraint:

Needs a host side fix and beefing up gtest to hit this same bug hopefully.

**Comment by tooniz** [comment]
I had to disable eager trace suite due to this hang. 

Please manually verify that the tests pass, and uncomment the following line in `.github/workflows/fast-dispatch-build-and-unit-tests.yaml` once the fix is in:
```
          # {name: eager trace tests, cmd: pytest $TT_METAL_HOME/tests/tt_eager/python_api_testing/trace_testing/ -vvv},
```

**Comment by kmabeeTT** [comment]
Hey @tooniz (if you see this) - Been debugging this further, and just found that the `DISPATCH_WAIT(count=expected_workers)` cmds that are normally between programs (trace disabled) are all missing (replaced with` DISPATCH_WAIT(count=0)` because `EnqueueProgramCommand` is created with `this->expected_num_workers_completed` (wasn't modified for trace feature) and in bypass mode (below logic) it is not incremented, instead `this->trace_ctx->num_completion_worker_cores` is incremented.

What's the reason for special handling of these worker-completions?

My ~2 line fix is below to use `this->trace_ctx->num_completion_worker_cores` when creating `EnqueueProgramCommand` when in bypass mode, so that `DISPATCH_WAIT()` is created with same number as when trace is disabled.  There were 2 programs in trace (1st uses 1 core, 2nd uses 64 cores) and this now allows the `DISPATCH_WAIT(1)` to be inserted before 2nd program (1 since prior program had 1 expected worker core).  Not yet sure if this is the right fix though (feels like it), but it works.

Open to thoughts.


```

void HWCommandQueue::enqueue_program() {

<snip>

    // KCM - Here is the fix, set env var to engage (illustrative purposes only)
    bool use_fix = tt::parse_env("USE_FIX", false);
    auto expected_workers = use_fix && this->manager.get_bypass_mode() ? this->trace_ctx->num_completion_worker_cores : this->expected_num_workers_completed;

    auto command = EnqueueProgramCommand(this->id, this->device, program, this->manager, expected_workers);
    this->enqueue_command(command, blocking);

    if (this->manager.get_bypass_mode()) {
        this->trace_ctx->num_completion_worker_cores += program.program_transfer_info.num_active_cores;
        this->trace_ctx->owned_buffer_pool.insert(this->trace_ctx->owned_buffer_pool.end(), program.kg_buffers.begin(), program.kg_buffers.end());
        this->trace_ctx->owned_buffer_pool.insert(this->trace_ctx->owned_buffer_pool.end(), program.owned_buffer_pool.begin(), program.owned_buffer_pool.end());
    } else {
        this->expected_num_workers_completed += program.program_transfer_info.num_active_cores;
    }
```

Update for this ticket/tracking - found a few other things along the way, will mention later.

**Comment by kmabeeTT** [comment]
Merged to fd-2/main at 18232d4f9 via PR https://github.com/tenstorrent/tt-metal/pull/7978 - marking closed.

**Comment by tooniz** [comment]
Hey @kmabeeTT you fix looks correct to me, once the program is being captuted (in bypass mode), `num_completion_worker_cores` from trace_ctx should be used instead, otherwise like you saw back to back programs' completions will be ignored since they get incremented on the trace_ctx `num_completion_worker_cores` only.

> same dispatch_wait(count) values of 1, 65 embedded in the cmds in the Trace are used in 2nd call to EnqueueTrace() like they were for 1st call. I think Trace intended to deal with this...somehow... (or maybe not)... need to think about it, or if there is some existing fix hiding somewhere...

Yes, trace does handle it. That the same values (1, 65) used in the trace assumes thatthe count starts from 0 at the start of the trace, that assumption relies on a count reset mechanism: in EnqueueTrace we first issue an wait to drain the device, but we set `clear` flag so `num_completion_worker_cores` is reset to 0 on device and we do the same on host.
